src/bluesky/simulators.py

```diff
@@ -104,3 +106,195 @@ async def check_limits_async(plan):
+            commands: the command name for the message as defined in bluesky Message Protocol, or a sequence if
+            more than one matches.
+            obj_name: the name property of the obj to match, can be None (the default) as not all messages have
+            a name.
+            predicate: if specified, then the predicate is used to select the handler rather than the object name.
+        """
+        if isinstance(commands, str):
+            commands = [commands]
+
+        self.message_handlers.insert(
+            0,
+            _MessageHandler(
+                lambda msg: msg.command in commands
+                and (
+                    (predicate and predicate(msg)) or (obj_name is None or (msg.obj and msg.obj.name == obj_name))
+                ),
+                handler,
+            ),
+        )
+
+    def add_read_handler(self, obj_name, value, dict_key="values"):
+        """
+        Convenience method to register a handler to return a result from a 'read' command.
+        Args:
+             value: The value to return
+             dict_key: The key in the result dictionary to populate with the value, by default this is 'values'
+
+        Example:
+            In order to exercise a plan containing the following:
+
+            >>> def trigger_and_return_pin_tip():
+            ...     yield from bps.trigger(pin_tip)
+            ...     tip_pos = yield from bps.rd(pin_tip.triggered_tip)
+            ...     return tip_pos
+
+
+            >>> sim = RunEngineSimulator()
+            >>> sim.add_read_handler("pin_tip_detection-triggered_tip", (8, 5))
+            >>> msgs = sim.simulate_plan(trigger_and_return_pin_tip())
+            >>> assert sim.return_value == (8, 5)
+        """
+
+        self.add_handler(lambda _: {dict_key: {"value": value}}, "read", obj_name)
+
+    def add_wait_handler(self, handler: Callable[[Msg], None], group: str = GROUP_ANY) -> None:
+        """Add a wait handler for a particular message
+        Args:
+            handler: a lambda that accepts a Msg, use this to execute any code that simulates something that's
+            supposed to complete when a group finishes
+            group: name of the group to wait for, default is GROUP_ANY which matches them all
+
+        Example:
+            In order to exercise the following plan:
+            >>> def close_shutter():
+            ...     yield from bps.abs_set(detector_motion.shutter, 1, group="my_group")
+            ...     yield from bps.wait("my_group")
+            ...     done = yield from bps.rd("detector_motion_shutter")
+            ...     assert done == 1
+
+            >>> sim = RunEngineSimulator()
+            ... def simulate_detector_motion():
+            ...     sim.add_read_handler("detector_motion_shutter", 1)
+            ...
+            ... sim.add_wait_handler(simulate_detector_motion, "my_group")
+            ... sim.simulate_plan(close_shutter())
+        """
+        self.message_handlers.append(
```

Comment:
I was thrown that this does not use `add_handler` which I see is because it does something that you can not do with the public API which makes me think that should be part of the public API.

Reply:
oh, reading the tests I understand now.  It seems very fiddily that you have to pass a function closing over a method on sim into sim.  Is there a way to have the handlers for document generating messages return the documents that should be emited?

Reply:
I think it's more likely that it was originally written before the other methods had reached their current form. I have made it delegate to `add_handler` but changed so that it has the same prepend behaviour to make it more consistent.